### PR TITLE
PIM-4945: rework the property adder

### PR DESCRIPTION
**Pull request by sumbobyboys**
| Q                 | A
| ----------------- | ---
| Added Specs       | N
| Added Behats      | N
| Travis CI is ok   | Y
| Changelog updated | Y

### CODE REVIEW

File: src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php [hunk 1]
```diff
@@ -24,4 +26,13 @@
...
     public function addData($object, $property, $data, array $options = []);
+
+    /**
+     * Get the adder for the following field
+     *
+     * @param string $field
+     *
+     * @return AdderInterface
+     */
+    public function getAdder($field);
```
Comment (nidup) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php]: imho, this is the job of the AdderRegistryInterface, i would prefer add a getAdder($property) in the registry (which encapsulate the resolution of attribute or field), here we mix two concerns, add a data in the property of an object, and get the relevant adder which is more a "internal concern"

Comment (sumbobyboys) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php]: You are 100% right. I will fix that right away

Comment (sumbobyboys) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php]: If we do that we will need to inject the attribute repository. Is it ok for you ?

File: src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 namespace Akeneo\Component\StorageUtils\Updater;
 
+use Pim\Component\Catalog\Updater\Adder\AdderInterface;
```
Comment (nidup) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyAdderInterface.php]: useless use

File: src/Akeneo/Component/StorageUtils/Updater/PropertyCopierInterface.php [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 namespace Akeneo\Component\StorageUtils\Updater;
 
+use Pim\Component\Catalog\Updater\Copier\CopierInterface;
```
Comment (nidup) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyCopierInterface.php]: useless use

File: src/Akeneo/Component/StorageUtils/Updater/PropertyRemoverInterface.php [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 namespace Akeneo\Component\StorageUtils\Updater;
 
+use Akeneo\Component\StorageUtils\Remover\RemoverInterface;
```
Comment (nidup) [inline src/Akeneo/Component/StorageUtils/Updater/PropertyRemoverInterface.php]: useless use

File: src/Akeneo/Component/StorageUtils/Updater/PropertySetterInterface.php [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 namespace Akeneo\Component\StorageUtils\Updater;
 
+use Pim\Component\Catalog\Updater\Setter\SetterInterface;
```
Comment (nidup) [inline src/Akeneo/Component/StorageUtils/Updater/PropertySetterInterface.php]: useless use

File: src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php [hunk 6]
```diff
@@ -19,6 +20,17 @@ class RemoverRegistry implements RemoverRegistryInterface
     /** @var FieldRemoverInterface[] priorized field removers */
     protected $fieldRemovers = [];
 
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $attributeRepository;
+
+    /**
+     * @param IdentifiableObjectRepositoryInterface $repository
+     */
+    public function __construct(IdentifiableObjectRepositoryInterface $repository)
```
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php]: BC break to indicate in the changelog

File: src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php [hunk 7]
```diff
@@ -19,6 +20,17 @@ class SetterRegistry implements SetterRegistryInterface
     /** @var FieldSetterInterface[] priorized field setters */
     protected $fieldSetters = [];
 
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $attributeRepository;
+
+    /**
+     * @param IdentifiableObjectRepositoryInterface $repository
+     */
+    public function __construct(IdentifiableObjectRepositoryInterface $repository)
```
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php]: BC break to indicate in the changelog

File: src/Pim/Component/Catalog/Updater/Copier/CopierRegistry.php [hunk 8]
```diff
@@ -19,6 +20,17 @@ class CopierRegistry implements CopierRegistryInterface
     /** @var FieldCopierInterface[] priorized field copiers */
     protected $fieldCopiers = [];
 
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $attributeRepository;
+
+    /**
+     * @param IdentifiableObjectRepositoryInterface $repository
+     */
+    public function __construct(IdentifiableObjectRepositoryInterface $repository)
```
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/Copier/CopierRegistry.php]: BC break to indicate in the changelog

File: src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php [hunk 9]
```diff
@@ -19,6 +20,17 @@ class AdderRegistry implements AdderRegistryInterface
     /** @var FieldAdderInterface[] priorized field adders */
     protected $fieldAdders = [];
 
+    /** @var IdentifiableObjectRepositoryInterface */
+    protected $attributeRepository;
+
+    /**
+     * @param IdentifiableObjectRepositoryInterface $repository
+     */
+    public function __construct(IdentifiableObjectRepositoryInterface $repository)
```
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php]: BC break to indicate in the changelog

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php [hunk 10]
```diff
@@ -74,6 +74,31 @@ public function getCategoriesByIds(array $categoriesIds = [])
     /**
      * {@inheritdoc}
      */
+    public function getCategoriesByCodes(array $categoriesCodes = [])
```
Comment (nidup) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: phpspec for this

File: src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php [hunk 11]
```diff
@@ -37,6 +49,21 @@ public function register(AdderInterface $adder)
     /**
      * {@inheritdoc}
      */
+    public function getAdder($property)
+    {
+        $attribute = $this->getAttribute($property);
+        if (null !== $attribute) {
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php]: ternary ?

**Comment by nidup** [comment]
fix comments then GTM

File: src/Pim/Bundle/CatalogBundle/Resources/config/updaters.yml [hunk 12]
```diff
@@ -143,6 +143,8 @@ services:
     # Setters
     pim_catalog.updater.setter.registry:
         class: %pim_catalog.updater.setter.registry.class%
+        arguments:
+            - '@pim_catalog.repository.cached_attribute'
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Resources/config/updaters.yml]: don't forget BC breaks for all of them

**Comment by skeleton** [comment]
baby comments then GTM

File: src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php (hunk 9 above)
Comment (sumbobyboys) [inline src/Pim/Component/Catalog/Updater/Adder/AdderRegistry.php]: done

File: src/Pim/Component/Catalog/Updater/Copier/CopierRegistry.php (hunk 8 above)
Comment (sumbobyboys) [inline src/Pim/Component/Catalog/Updater/Copier/CopierRegistry.php]: done

File: src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php (hunk 7 above)
Comment (sumbobyboys) [inline src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php]: done

File: src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php (hunk 6 above)
Comment (sumbobyboys) [inline src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php]: done

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php (hunk 10 above)
Comment (sumbobyboys) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: We don't have any phpspec for this repository. Do I create One ?

**Comment by sumbobyboys** [comment]
Thanks !

File: src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php [hunk 13]
```diff
@@ -74,6 +74,31 @@ public function getCategoriesByIds(array $categoriesIds = [])
...
+    {
+        if (empty($categoriesCodes)) {
+            return new ArrayCollection();
+        }
+
+        $meta = $this->getClassMetadata();
+        $config = $this->listener->getConfiguration($this->_em, $meta->name);
```
Comment (phaseinducer) [inline src/Akeneo/Bundle/ClassificationBundle/Doctrine/ORM/Repository/CategoryRepository.php]: align please

File: src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php [hunk 14]
```diff
@@ -37,6 +49,21 @@ public function register(RemoverInterface $remover)
     /**
      * {@inheritdoc}
      */
+    public function getRemover($property)
+    {
+        $attribute = $this->getAttribute($property);
+        if (null !== $attribute) {
```
Comment (phaseinducer) [inline src/Pim/Component/Catalog/Updater/Remover/RemoverRegistry.php]: ternary

File: src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php [hunk 15]
```diff
@@ -37,6 +49,21 @@ public function register(SetterInterface $setter)
     /**
      * {@inheritdoc}
      */
+    public function getSetter($property)
+    {
+        $attribute = $this->getAttribute($property);
+        if (null !== $attribute) {
```
Comment (phaseinducer) [inline src/Pim/Component/Catalog/Updater/Setter/SetterRegistry.php]: same

**Comment by phaseinducer** [comment]
little comments & GTM

**Pull request merged**